### PR TITLE
post-restore: add optional level for post-restore operations

### LINKED ISSUE #403: Add an intermediate "optional" level for post-restore operations (CHECKSUM, ANALYZE)
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently if CHECKSUM and ANALYZE failed we will abort the entire task. This may not be desirable as sometimes the failure is spurious due to timeout or maybe the user just don't care.

**Describe the feature you'd like:**

Change the `post-restore.checksum` and `.analyze` options from `false` and `true` to:

* `"off"` — disable entirely, same as `false`
* `"optional"` — enable, but on failure just print a WARN (or ERROR?) level log without killing the task
* `"required"` — enable, and kill the task on failure, same as `true`.

The default of `checksum` is changed from `true` to `"required"`, and the default of `analyze` is changed from `true` to `"optional"`.

For backward compatibility the boolean options are continued to be accepted forever.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Optimization:**

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
add optional level for post-restore operations close #403.

### What is changed and how it works?
Add operation level: 'off', 'optional', 'required' for post-restore operations checksum and analyze.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
 - No code

Side effects

Related changes

 - Need to update the documentation

Release Note
- Add an intermediate "optional" level for post-restore operations

### CODE REVIEW

File: lightning/config/config.go [hunk 1]
```diff
@@ -127,12 +127,55 @@ type Lightning struct {
...
+type PostOpLevel int
+
+const (
+	OpLevelOff PostOpLevel = iota
+	OPLevelOptional
+	OpLevelRequired
+)
+
+func (t *PostOpLevel) UnmarshalText(text []byte) error {
+	switch strings.ToLower(string(text)) {
```
Comment (kennytm) [inline lightning/config/config.go:139]: i don't think you can implement `UnmarshalText` this way. The `text` can be `"required"` or `'required'` or `"""""required"""""`. Better implement `UnmarshalTOML`.

**Review by kennytm — COMMENTED** [review]
unit test failed.

File: lightning/config/config.go (hunk 1 above)
Comment (glorv) [reply lightning/config/config.go:139]: is there a interface `UnmarshalTOML` in the toml or encoding package? The toml souce code show table it use the `encoding.TextUnmarshaler` to decode? 
Shall we do a strings.Trim(text, `'"`) here?

Comment (kennytm) [reply lightning/config/config.go]: >  Shall we do a strings.Trim(text, `'"`) here?

No

> is there a interface UnmarshalTOML in the toml or encoding package? 

https://pkg.go.dev/github.com/BurntSushi/toml#example-package-UnmarshalTOML

Comment (glorv) [reply lightning/config/config.go]: fixed

File: lightning/config/config.go [hunk 2]
```diff
@@ -127,12 +127,68 @@ type Lightning struct {
...
+func (t *PostOpLevel) UnmarshalTOML(v interface{}) error {
+	switch v.(type) {
+	case bool:
+		if v.(bool) {
+			*t = OpLevelRequired
+		} else {
+			*t = OpLevelOff
+		}
+	case string:
+		return t.FromStringValue(v.(string))
```
Comment (kennytm) [inline lightning/config/config.go:147]:
```suggestion
	switch val := v.(type) {
	case bool:
		if val {
			*t = OpLevelRequired
		} else {
			*t = OpLevelOff
		}
	case string:
		return t.FromStringValue(val)
```

File: lightning/config/config.go [hunk 3]
```diff
@@ -127,12 +127,68 @@ type Lightning struct {
 	CheckRequirements bool `toml:"check-requirements" json:"check-requirements"`
 }
 
+type PostOpLevel int
+
+const (
+	OpLevelOff PostOpLevel = iota
+	OPLevelOptional
```
Comment (kennytm) [inline lightning/config/config.go:134]:
```suggestion
	OpLevelOptional
```

**Review by kennytm — COMMENTED** [review]
rest LGTM, and you need to implement `UnmarshalText`.

```
[2020-10-15T03:04:59.361Z] FAIL: lightning_test.go:191: lightningServerSuite.TestGetDeleteTask
[2020-10-15T03:04:59.361Z] 
[2020-10-15T03:04:59.361Z] lightning_test.go:271:
[2020-10-15T03:04:59.361Z]     c.Assert(err, IsNil)
[2020-10-15T03:04:59.361Z] ... value *json.UnmarshalTypeError = &json.UnmarshalTypeError{Value:"string", Type:(*reflect.rtype)(0x1cbd640), Offset:1342, Struct:"PostRestore", Field:"post-restore.checksum"} ("json: cannot unmarshal string into Go struct field PostRestore.post-restore.checksum of type config.PostOpLevel")
```

**Comment by kennytm** [comment]
lgtm

**Comment by overvenus** [comment]
LGTM

**Review by kennytm — APPROVED** [review]

**Pull request merged**